### ISSUE TITLE
@ sign breaks importing of index

### ISSUE THREAD

**Issue by corroded** [karmi/retire#93]
Hi,

We think that there's a problem with the importing when it is encountering @-signs. We thought we fixed this problem before by adding `encoding: utf8` in our database.yml but now it is appearing when encountering @-signs:

```
Artist.import
[ERROR] 400 Bad Request:{"error":"Unexpected character ('@' (code 64)): expected a valid value (number, String, array, object, 'true', 'false' or 'null')\n at [Source: [B@17ce10d; line: 1, column: 109120]"}, retrying (1)...
[ERROR] 400 Bad Request:{"error":"Unexpected character ('@' (code 64)): expected a valid value (number, String, array, object, 'true', 'false' or 'null')\n at [Source: [B@8e261b; line: 1, column: 109120]"}, retrying (2)...
[ERROR] 400 Bad Request:{"error":"Unexpected character ('@' (code 64)): expected a valid value (number, String, array, object, 'true', 'false' or 'null')\n at [Source: [B@d65732; line: 1, column: 109120]"}, retrying (3)...
[ERROR] 400 Bad Request:{"error":"Unexpected character ('@' (code 64)): expected a valid value (number, String, array, object, 'true', 'false' or 'null')\n at [Source: [B@14cc6; line: 1, column: 109120]"}, retrying (4)...
[ERROR] 400 Bad Request:{"error":"Unexpected character ('@' (code 64)): expected a valid value (number, String, array, object, 'true', 'false' or 'null')\n at [Source: [B@174343; line: 1, column: 109120]"}, retrying (5)...
[ERROR] Too many exceptions occured, giving up...
Response: {"error":"Unexpected character ('@' (code 64)): expected a valid value (number, String, array, object, 'true', 'false' or 'null')\n at [Source: [B@1f6913a; line: 1, column: 109120]"}
RestClient::BadRequest: 400 Bad Request
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/rest-client-1.6.7/lib/restclient/abstract_response.rb:48:in `return!'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/rest-client-1.6.7/lib/restclient/request.rb:230:in `process_result'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/rest-client-1.6.7/lib/restclient/request.rb:178:in `transmit'
    from /Users/dev/.rvm/rubies/ree-1.8.7-2011.03/lib/ruby/1.8/net/http.rb:543:in `start'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/rest-client-1.6.7/lib/restclient/request.rb:172:in `transmit'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/rest-client-1.6.7/lib/restclient/request.rb:64:in `execute'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/rest-client-1.6.7/lib/restclient/request.rb:33:in `execute'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/rest-client-1.6.7/lib/restclient.rb:72:in `post'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/tire-0.3.2/lib/tire/client.rb:10:in `post'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/tire-0.3.2/lib/tire/index.rb:86:in `bulk_store'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/tire-0.3.2/lib/tire/index.rb:112:in `import'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/tire-0.3.2/lib/tire/model/import.rb:17:in `import'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/tire-0.3.2/lib/tire/model/search.rb:249:in `__send__'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/tire-0.3.2/lib/tire/model/search.rb:249:in `import'
    from (irb):8
    from :0
```

also there's a redundant utf-8 problem:

```
Song.limit(10000).import
ArgumentError: redundant UTF-8 sequence
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/activesupport-3.0.9/lib/active_support/json/encoding.rb:130:in `unpack'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/activesupport-3.0.9/lib/active_support/json/encoding.rb:130:in `escape'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/activesupport-3.0.9/lib/active_support/json/encoding.rb:125:in `gsub'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/activesupport-3.0.9/lib/active_support/json/encoding.rb:125:in `escape'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/activesupport-3.0.9/lib/active_support/json/encoding.rb:68:in `escape'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/activesupport-3.0.9/lib/active_support/json/encoding.rb:173:in `encode_json'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/activesupport-3.0.9/lib/active_support/json/encoding.rb:47:in `encode'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/activesupport-3.0.9/lib/active_support/json/encoding.rb:77:in `check_for_circular_references'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/activesupport-3.0.9/lib/active_support/json/encoding.rb:45:in `encode'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/activesupport-3.0.9/lib/active_support/json/encoding.rb:247:in `encode_json'
    from (irb):12:in `map'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/activesupport-3.0.9/lib/active_support/json/encoding.rb:247:in `each'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/activesupport-3.0.9/lib/active_support/json/encoding.rb:247:in `map'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/activesupport-3.0.9/lib/active_support/json/encoding.rb:247:in `encode_json'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/activesupport-3.0.9/lib/active_support/json/encoding.rb:47:in `encode'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/activesupport-3.0.9/lib/active_support/json/encoding.rb:77:in `check_for_circular_references'
... 4 levels...
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/tire-0.3.2/lib/tire/index.rb:282:in `convert_document_to_json'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/tire-0.3.2/lib/tire/index.rb:77:in `bulk_store'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/activerecord-3.0.9/lib/active_record/relation.rb:13:in `map'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/activerecord-3.0.9/lib/active_record/relation.rb:13:in `__send__'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/activerecord-3.0.9/lib/active_record/relation.rb:13:in `map'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/tire-0.3.2/lib/tire/index.rb:69:in `bulk_store'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/tire-0.3.2/lib/tire/index.rb:112:in `import'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/tire-0.3.2/lib/tire/model/import.rb:17:in `import'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/tire-0.3.2/lib/tire/model/search.rb:249:in `__send__'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/tire-0.3.2/lib/tire/model/search.rb:249:in `import'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/activerecord-3.0.9/lib/active_record/relation.rb:370:in `send'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/activerecord-3.0.9/lib/active_record/relation.rb:370:in `method_missing'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/activerecord-3.0.9/lib/active_record/relation.rb:125:in `scoping'
    from /Users/dev/.rvm/gems/ree-1.8.7-2011.03@ourapp2/gems/activerecord-3.0.9/lib/active_record/relation.rb:370:in `method_missing'
    from (irb):12
```

**Comment by karmi** [comment]
Tire, by itself has no problem with importing @ signs:

``` ruby
    require 'rubygems'
    require 'tire'

    documents = [ {:id => 1, :body => 'user@foo.bar'}, {:id => 2, :body => 'foo bar'} ]

    Tire.index 'at_signs_index' do
      delete
      create
      import documents

      refresh
    end

    s = Tire.search('at_signs_index') { query { string 'user@foo.bar' } }

    p s.results.first
```

There's something else going on in your codebase, which makes the JSON serialization fail. Could you isolate this so I can try to recreate your problem?

**Comment by ramontayag** [comment]
@karmi, you're right. I'm working with @corroded on this and it **seems** that some of the string in our data has some bytecode (?) / bytedata characters. I don't really know anything about this so I'll have to research more. I guess that's why we see the `code 64` in the error. We'll send updates!

**Comment by ramontayag** [comment]
I can't seem to reproduce it. Importing is working fine now. It shames me to say this, but the problem may have been us not restarting the rails console (I had moved to a new workstation without `encoding` specified in the database.yml. I changed it, called `reload!` in the console, but apparently, `reload!` wasn't enough) after setting the encoding in database.yml. Because after I restarted the console for other reasons, and started importing, things worked. Because of this, this issue is invalid.

Thanks @karmi! You've been a great help!
